docs/DevGuide/CompilerLinkerErrors.md

```diff
@@ -0,0 +1,115 @@
+brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial,
+(IndexType)0> > >
+random_unit_normal<DataVector>(gsl::not_null<std::mersenne_twister_engine<
+unsigned long, 32ul, 624ul, 397ul, 31ul, 2567483615ul, 11ul, 4294967295ul, 7ul,
+2636928640ul, 15ul, 4022730752ul, 18ul, 1812433253ul>*>, Tensor<DataVector,
+brigand::list<brigand::integral_constant<int, 1>,
+brigand::integral_constant<int, 1> >,
+brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial,
+(IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial,
+(IndexType)0> > > const&)
+```
+In order to make this easier to read, it is recommended to run the code through
+ClangFormat. This can be done by copying the linker output into an empty `cpp`
+file and running `clang-format -i EMPTY_FILE_WITH_LINKER_OUTPUT`. Keep in mind
+that ClangFormat will only work if you have only copied the part of the linker
+output that resembles valid C++. Doing so in this case gives:
+\code{.cpp}
+Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1> >,
+       brigand::list<Tensor_detail::TensorIndexType<
+           3ul, (UpLo)0, Frame::Inertial, (IndexType)0> > >
+random_unit_normal<DataVector>(
+    gsl::not_null<std::mersenne_twister_engine<
+        unsigned long, 32ul, 624ul, 397ul, 31ul, 2567483615ul, 11ul,
+        4294967295ul, 7ul, 2636928640ul, 15ul, 4022730752ul, 18ul,
+        1812433253ul>*>,
+    Tensor<DataVector,
+           brigand::list<brigand::integral_constant<int, 1>,
+                         brigand::integral_constant<int, 1> >,
+           brigand::list<
+               Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial,
+                                              (IndexType)0>,
+               Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial,
+                                              (IndexType)0> > > const&);
+\endcode
+We see that it is the function `random_unit_normal` that isn't found. Now we can
+search in the code base where than function is defined. Doing a
+`git grep "random_unit_normal"` points to
+`tests/Unit/Helpers/DataStructures/RandomUnitNormal.?pp`. Looking
+at `tests/Unit/Helpers/DataStructures/RandomUnitNormal.hpp` we see that the
+function `random_unit_normal` is declared there, and looking in the
+corresponding `cpp` we see `random_unit_normal` is instantiated in the source
+file. Opening up `tests/Unit/Helpers/DataStructures/CMakeLists.txt` we see that
+the library name is `DataStructuresHelpers`. Thus, linking
+`Test_GeneralizedHarmonic` against `DataStructuresHelpers` will resolve our
+error. To link against a library, you must add it to the
+`target_link_libraries`, or the last argument passed to `add_test_library`.
+
+If `random_unit_normal` had been defined in the header file, then the error
+would've indicated that we did not include the header (or `tpp`) file into
+`Test_UpwindFlux.cpp`, and so the compiler could not generate an
+instantiation.
```

Comment:
maybe add a checklist (at the end or before the detailed explanation). Something like:

- Identify target (file included in library or executable)
- Identify missing source definition (usually a function or static variable)
- Find source declaration and definition in code.
- If source definition is in cpp file, make sure it is in list of sources in CMakeLists.txt and the corresponding library is linked against by the target.  If the source is a template, make sure the specific instantiation exists.
- If source definition is in a tpp file, make sure the tpp file in included in the target file.
- If source definition is in a hpp file, make sure the specific instantiation is possible (e.g. not forbidden via a `Requires`)